File: backend/StockAnalyzer.Api/Services/StockDataService.cs
using System.Net.Http.Json;
using System.Text.Json;
using StockAnalyzer.Api.Models;

namespace StockAnalyzer.Api.Services;

public class StockDataService : IStockDataService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _configuration;
    private readonly ILogger<StockDataService> _logger;
    private readonly string _apiKey;
    private readonly string _apiUrl;

    public StockDataService(
        HttpClient httpClient,
        IConfiguration configuration,
        ILogger<StockDataService> logger)
    {
        _httpClient = httpClient;
        _configuration = configuration;
        _logger = logger;
        
        // Get API key from environment variable or configuration
        _apiKey = Environment.GetEnvironmentVariable("TWELVE_DATA_API_KEY") 
            ?? _configuration["TwelveData:ApiKey"] 
            ?? "";
        
        if (string.IsNullOrEmpty(_apiKey))
        {
            _logger.LogWarning("Twelve Data API key not found. Set TWELVE_DATA_API_KEY environment variable or configure in appsettings.json");
        }
        else
        {
            _logger.LogInformation("Twelve Data API key loaded successfully (length: {Length})", _apiKey.Length);
        }
        
        _apiUrl = _configuration["TwelveData:ApiUrl"] ?? "https://api.twelvedata.com";
    }

    public async Task<List<StockSearchResult>> SearchStocksAsync(string query)
    {
        try
        {
            var url = $"{_apiUrl}/symbol_search?symbol={Uri.EscapeDataString(query)}&apikey={_apiKey}";
            var response = await _httpClient.GetAsync(url);
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            var searchResults = new List<StockSearchResult>();

            if (result.TryGetProperty("data", out var dataElement) && dataElement.ValueKind == JsonValueKind.Array)
            {
                foreach (var item in dataElement.EnumerateArray())
                {
                    searchResults.Add(new StockSearchResult
                    {
                        Symbol = item.TryGetProperty("symbol", out var symbol) ? symbol.GetString() ?? "" : "",
                        Name = item.TryGetProperty("instrument_name", out var name) ? name.GetString() ?? "" : "",
                        Exchange = item.TryGetProperty("exchange", out var exchange) ? exchange.GetString() ?? "" : "",
                        Type = item.TryGetProperty("instrument_type", out var type) ? type.GetString() ?? "" : ""
                    });
                }
            }

            return searchResults;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error searching stocks for query: {Query}", query);
            throw;
        }
    }

    public async Task<StockQuote?> GetStockQuoteAsync(string symbol)
    {
        try
        {
            var url = $"{_apiUrl}/quote?symbol={Uri.EscapeDataString(symbol)}&apikey={_apiKey}";
            var response = await _httpClient.GetAsync(url);
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            
            if (result.TryGetProperty("status", out var status) && status.GetString() == "error")
            {
                return null;
            }

            var price = result.TryGetProperty("close", out var close) ? decimal.Parse(close.GetString() ?? "0") : 0;
            var previousClose = result.TryGetProperty("previous_close", out var prevClose) ? decimal.Parse(prevClose.GetString() ?? "0") : price;
            var change = result.TryGetProperty("change", out var changeProp) ? decimal.Parse(changeProp.GetString() ?? "0") : (price - previousClose);
            var changePercent = result.TryGetProperty("percent_change", out var changePct) ? decimal.Parse(changePct.GetString() ?? "0") : (previousClose != 0 ? (change / previousClose) * 100 : 0);
            var volume = result.TryGetProperty("volume", out var vol) ? long.Parse(vol.GetString() ?? "0") : 0;

            return new StockQuote
            {
                Symbol = result.TryGetProperty("symbol", out var sym) ? sym.GetString() ?? symbol : symbol,
                Price = price,
                Change = change,
                ChangePercent = changePercent,
                Volume = volume,
                Timestamp = DateTime.UtcNow
            };
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting stock quote for symbol: {Symbol}", symbol);
            throw;
        }
    }

    public async Task<StockDetails?> GetStockDetailsAsync(string symbol)
    {
        try
        {
            // Fetch both quote and profile data
            var quoteTask = GetStockQuoteAsync(symbol);
            var profileUrl = $"{_apiUrl}/profile?symbol={Uri.EscapeDataString(symbol)}&apikey={_apiKey}";
            var profileResponse = await _httpClient.GetAsync(profileUrl);
            
            var quote = await quoteTask;
            if (quote == null)
            {
                return null;
            }

            var details = new StockDetails
            {
                Symbol = quote.Symbol,
                CurrentPrice = quote.Price,
                Change = quote.Change,
                ChangePercent = quote.ChangePercent,
                Volume = quote.Volume
            };

            if (profileResponse.IsSuccessStatusCode)
            {
                var profile = await profileResponse.Content.ReadFromJsonAsync<JsonElement>();
                
                // Check if the response contains an error (even with 200 status)
                if (profile.TryGetProperty("status", out var status) && status.GetString() == "error")
                {
                    var errorMessage = profile.TryGetProperty("message", out var msg) ? msg.GetString() : "Unknown error";
                    _logger.LogWarning("Profile API returned error for symbol: {Symbol}. Message: {Message}", symbol, errorMessage);
                    // Profile endpoint requires paid plan - skip parsing profile data
                }
                else
                {
                    _logger.LogInformation("Profile data parsed successfully for symbol: {Symbol}", symbol);
                    
                    // Basic information
                details.Name = profile.TryGetProperty("name", out var name) ? name.GetString() ?? "" : "";
                details.Exchange = profile.TryGetProperty("exchange", out var exchange) ? exchange.GetString() ?? "" : "";
                details.Sector = profile.TryGetProperty("sector", out var sector) ? sector.GetString() : null;
                details.Industry = profile.TryGetProperty("industry", out var industry) ? industry.GetString() : null;
                details.Description = profile.TryGetProperty("description", out var desc) ? desc.GetString() : null;
                
                // Market capitalization
                if (profile.TryGetProperty("market_capitalization", out var mcap))
                {
                    if (mcap.ValueKind == JsonValueKind.String)
                    {
                        var mcapStr = mcap.GetString();
                        if (!string.IsNullOrEmpty(mcapStr) && decimal.TryParse(mcapStr, out var mcapValue))
                        {
                            details.MarketCap = mcapValue;
                        }
                    }
                    else if (mcap.ValueKind == JsonValueKind.Number)
                    {
                        details.MarketCap = mcap.GetDecimal();
                    }
                }
                
                // P/E Ratio
                if (profile.TryGetProperty("pe_ratio", out var pe))
                {
                    if (pe.ValueKind == JsonValueKind.String)
                    {
                        var peStr = pe.GetString();
                        if (!string.IsNullOrEmpty(peStr) && decimal.TryParse(peStr, out var peValue))
                        {
                            details.PeRatio = peValue;
                        }
                    }
                    else if (pe.ValueKind == JsonValueKind.Number)
                    {
                        details.PeRatio = pe.GetDecimal();
                    }
                }
                
                // Dividend Yield
                if (profile.TryGetProperty("dividend_yield", out var divYield))
                {
                    if (divYield.ValueKind == JsonValueKind.String)
                    {
                        var divStr = divYield.GetString();
                        if (!string.IsNullOrEmpty(divStr) && decimal.TryParse(divStr, out var divValue))
                        {
                            details.DividendYield = divValue;
                        }
                    }
                    else if (divYield.ValueKind == JsonValueKind.Number)
                    {
                        details.DividendYield = divYield.GetDecimal();
                    }
                }
                
                // 52 Week High
                if (profile.TryGetProperty("52_week_high", out var high52) || profile.TryGetProperty("fifty_two_week_high", out high52))
                {
                    if (high52.ValueKind == JsonValueKind.String)
                    {
                        var highStr = high52.GetString();
                        if (!string.IsNullOrEmpty(highStr) && decimal.TryParse(highStr, out var highValue))
                        {
                            details.High52Week = highValue;
                        }
                    }
                    else if (high52.ValueKind == JsonValueKind.Number)
                    {
                        details.High52Week = high52.GetDecimal();
                    }
                }
                
                // 52 Week Low
                if (profile.TryGetProperty("52_week_low", out var low52) || profile.TryGetProperty("fifty_two_week_low", out low52))
                {
                    if (low52.ValueKind == JsonValueKind.String)
                    {
                        var lowStr = low52.GetString();
                        if (!string.IsNullOrEmpty(lowStr) && decimal.TryParse(lowStr, out var lowValue))
                        {
                            details.Low52Week = lowValue;
                        }
                    }
                    else if (low52.ValueKind == JsonValueKind.Number)
                    {
                        details.Low52Week = low52.GetDecimal();
                    }
                }
                
                // Average Volume
                if (profile.TryGetProperty("average_volume", out var avgVol))
                {
                    if (avgVol.ValueKind == JsonValueKind.String)
                    {
                        var avgVolStr = avgVol.GetString();
                        if (!string.IsNullOrEmpty(avgVolStr) && long.TryParse(avgVolStr, out var avgVolValue))
                        {
                            details.AverageVolume = avgVolValue;
                        }
                    }
                    else if (avgVol.ValueKind == JsonValueKind.Number)
                    {
                        details.AverageVolume = (long)avgVol.GetInt64();
                    }
                }
                }
            }
            else
            {
                var errorContent = await profileResponse.Content.ReadAsStringAsync();
                _logger.LogWarning("Profile API call failed for symbol: {Symbol}. Status: {Status}, Response: {Response}", 
                    symbol, profileResponse.StatusCode, errorContent);
            }
            
            // Try to get 52-week high/low from historical data if profile is not available
            // Note: Free plan may limit historical data, so try smaller ranges first
            if (details.High52Week == null || details.Low52Week == null)
            {
                _logger.LogInformation("Attempting to get 52-week high/low from historical data for symbol: {Symbol}", symbol);
                try
                {
                    // Try with available data (free plan typically allows up to 30 days)
                    // We'll use whatever data is available to calculate the range
                    var historicalData = await GetHistoricalDataAsync(symbol, "1day", 30);
                    _logger.LogInformation("Historical data retrieved. Count: {Count}", historicalData?.Count ?? 0);
                    
                    if (historicalData != null && historicalData.Count > 0)
                    {
                        details.High52Week = historicalData.Max(d => d.High);
                        details.Low52Week = historicalData.Min(d => d.Low);
                        _logger.LogInformation("Calculated high/low from historical data for symbol: {Symbol}. High: {High}, Low: {Low} (based on {Days} days)", 
                            symbol, details.High52Week, details.Low52Week, historicalData.Count);
                    }
                    else
                    {
                        _logger.LogWarning("Historical data is null or empty for symbol: {Symbol}. Free plan may limit historical data access.", symbol);
                    }
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, "Failed to get 52-week high/low from historical data for symbol: {Symbol}", symbol);
                }
            }

            return details;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting stock details for symbol: {Symbol}", symbol);
            throw;
        }
    }

    public async Task<List<StockQuote>> GetMultipleStockQuotesAsync(List<string> symbols)
    {
        try
        {
            var symbolsParam = string.Join(",", symbols);
            var url = $"{_apiUrl}/quote?symbol={Uri.EscapeDataString(symbolsParam)}&apikey={_apiKey}";
            var response = await _httpClient.GetAsync(url);
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            var quotes = new List<StockQuote>();

            // Twelve Data returns multiple quotes as an object with symbol keys
            if (result.ValueKind == JsonValueKind.Object)
            {
                foreach (var symbol in symbols)
                {
                    if (result.TryGetProperty(symbol, out var quoteElement))
                    {
                        var price = quoteElement.TryGetProperty("close", out var close) ? decimal.Parse(close.GetString() ?? "0") : 0;
                        var previousClose = quoteElement.TryGetProperty("previous_close", out var prevClose) ? decimal.Parse(prevClose.GetString() ?? "0") : price;
                        var change = quoteElement.TryGetProperty("change", out var changeProp) ? decimal.Parse(changeProp.GetString() ?? "0") : (price - previousClose);
                        var changePercent = quoteElement.TryGetProperty("percent_change", out var changePct) ? decimal.Parse(changePct.GetString() ?? "0") : (previousClose != 0 ? (change / previousClose) * 100 : 0);
                        var volume = quoteElement.TryGetProperty("volume", out var vol) ? long.Parse(vol.GetString() ?? "0") : 0;

                        quotes.Add(new StockQuote
                        {
                            Symbol = symbol,
                            Price = price,
                            Change = change,
                            ChangePercent = changePercent,
                            Volume = volume,
                            Timestamp = DateTime.UtcNow
                        });
                    }
                }
            }

            return quotes;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting multiple stock quotes");
            throw;
        }
    }

    public async Task<List<StockHistoricalData>> GetHistoricalDataAsync(string symbol, string interval = "1day", int? outputSize = null)
    {
        try
        {
            // Default to 30 days if outputSize not specified
            if (!outputSize.HasValue)
            {
                outputSize = interval switch
                {
                    "1day" => 30,
                    "1week" => 12,
                    "1month" => 12,
                    _ => 30
                };
            }

            var url = $"{_apiUrl}/time_series?symbol={Uri.EscapeDataString(symbol)}&interval={interval}&outputsize={outputSize.Value}&apikey={_apiKey}";

            var response = await _httpClient.GetAsync(url);
            response.EnsureSuccessStatusCode();

            var result = await response.Content.ReadFromJsonAsync<JsonElement>();
            
            if (result.TryGetProperty("status", out var status) && status.GetString() == "error")
            {
                var errorMessage = result.TryGetProperty("message", out var msg) ? msg.GetString() : "Unknown error";
                _logger.LogWarning("Historical data API returned error for symbol: {Symbol}. Message: {Message}", symbol, errorMessage);
                return new List<StockHistoricalData>();
            }

            var historicalData = new List<StockHistoricalData>();

            if (result.TryGetProperty("values", out var values) && values.ValueKind == JsonValueKind.Array)
            {
                foreach (var value in values.EnumerateArray())
                {
                    if (DateTime.TryParse(value.TryGetProperty("datetime", out var dt) ? dt.GetString() : null, out var date))
                    {
                        historicalData.Add(new StockHistoricalData
                        {
                            Date = date,
                            Open = value.TryGetProperty("open", out var open) ? decimal.Parse(open.GetString() ?? "0") : 0,
                            High = value.TryGetProperty("high", out var high) ? decimal.Parse(high.GetString() ?? "0") : 0,
                            Low = value.TryGetProperty("low", out var low) ? decimal.Parse(low.GetString() ?? "0") : 0,
                            Close = value.TryGetProperty("close", out var close) ? decimal.Parse(close.GetString() ?? "0") : 0,
                            Volume = value.TryGetProperty("volume", out var vol) ? long.Parse(vol.GetString() ?? "0") : 0
                        });
                    }
                }
            }

            // Return data sorted by date (oldest first for chart)
            return historicalData.OrderBy(d => d.Date).ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error getting historical data for symbol: {Symbol}", symbol);
            throw;
        }
    }
}

File: backend/services/stock_data_service.py
"""Twelve Data API client for stock search, quotes, details, and historical data."""
import logging
import asyncio
from datetime import UTC, datetime
from typing import Any
from urllib.parse import quote_plus

import httpx

from config import TWELVE_DATA_API_KEY, TWELVE_DATA_API_URL
from models.stock_models import (
    StockDetails,
    StockHistoricalData,
    StockQuote,
    StockSearchResult,
)

logger = logging.getLogger(__name__)
DEFAULT_TIMEOUT = httpx.Timeout(10.0)


class StockDataConfigurationError(RuntimeError):
    pass


def _require_api_key() -> None:
    if not TWELVE_DATA_API_KEY:
        raise StockDataConfigurationError("TWELVE_DATA_API_KEY is not configured")


async def search_stocks(query: str) -> list[StockSearchResult]:
    if not query or not query.strip():
        return []
    _require_api_key()
    url = f"{TWELVE_DATA_API_URL}/symbol_search?symbol={quote_plus(query.strip())}&apikey={TWELVE_DATA_API_KEY}"
    async with httpx.AsyncClient(timeout=DEFAULT_TIMEOUT) as client:
        resp = await client.get(url)
        resp.raise_for_status()
    data = resp.json()
    results = []
    for item in data.get("data", [])[:5]:
        results.append(
            StockSearchResult(
                symbol=item.get("symbol", ""),
                name=item.get("instrument_name", ""),
                exchange=item.get("exchange", ""),
                type=item.get("instrument_type", ""),
            )
        )
    return results


def _parse_decimal(val: Any, default: float = 0.0) -> float:
    if val is None:
        return default
    if isinstance(val, (int, float)):
        return float(val)
    try:
        return float(val)
    except (TypeError, ValueError):
        return default


def _parse_int(val: Any, default: int = 0) -> int:
    if val is None:
        return default
    if isinstance(val, int):
        return val
    try:
        return int(float(val))
    except (TypeError, ValueError):
        return default


def _needs_name_fallback(symbol: str, name: str | None) -> bool:
    normalized_symbol = (symbol or "").strip().upper()
    normalized_name = (name or "").strip().upper()
    return not normalized_name or normalized_name == normalized_symbol


async def get_stock_quote(symbol: str, client: httpx.AsyncClient | None = None) -> StockQuote | None:
    _require_api_key()
    url = f"{TWELVE_DATA_API_URL}/quote?symbol={quote_plus(symbol)}&apikey={TWELVE_DATA_API_KEY}"
    if client:
        resp = await client.get(url)
        resp.raise_for_status()
    else:
        async with httpx.AsyncClient(timeout=DEFAULT_TIMEOUT) as owned_client:
            resp = await owned_client.get(url)
            resp.raise_for_status()
    result = resp.json()
    if result.get("status") == "error":
        return None
    price = _parse_decimal(result.get("close"), 0)
    prev_close = _parse_decimal(result.get("previous_close"), price)
    change = _parse_decimal(result.get("change"), price - prev_close)
    change_pct = _parse_decimal(
        result.get("percent_change"),
        (change / prev_close * 100) if prev_close else 0,
    )
    volume = _parse_int(result.get("volume"), 0)
    return StockQuote(
        symbol=result.get("symbol", symbol),
        price=price,
        change=change,
        change_percent=change_pct,
        volume=volume,
        timestamp=datetime.now(UTC),
    )


async def get_stock_details(symbol: str) -> StockDetails | None:
    quote = await get_stock_quote(symbol)
    if not quote:
        return None
    details = StockDetails(
        symbol=quote.symbol,
        current_price=quote.price,
        change=quote.change,
        change_percent=quote.change_percent,
        volume=quote.volume,
    )
    profile_url = f"{TWELVE_DATA_API_URL}/profile?symbol={quote_plus(symbol)}&apikey={TWELVE_DATA_API_KEY}"
    async with httpx.AsyncClient(timeout=DEFAULT_TIMEOUT) as client:
        profile_resp = await client.get(profile_url)
    if profile_resp.is_success:
        profile = profile_resp.json()
        if profile.get("status") == "error":
            logger.warning(
                "Profile API returned error for symbol %s: %s",
                symbol,
                profile.get("message", "Unknown"),
            )
        else:
            details.name = profile.get("name", "")
            details.exchange = profile.get("exchange", "")
            details.sector = profile.get("sector")
            details.industry = profile.get("industry")
            details.description = profile.get("description")
            details.market_cap = _parse_decimal(profile.get("market_capitalization")) or None
            details.pe_ratio = _parse_decimal(profile.get("pe_ratio")) or None
            details.dividend_yield = _parse_decimal(profile.get("dividend_yield")) or None
            details.high_52_week = _parse_decimal(profile.get("52_week_high") or profile.get("fifty_two_week_high")) or None
            details.low_52_week = _parse_decimal(profile.get("52_week_low") or profile.get("fifty_two_week_low")) or None
            details.average_volume = _parse_int(profile.get("average_volume")) or None
    if _needs_name_fallback(symbol, details.name):
        try:
            search_results = await search_stocks(symbol)
            exact_match = next(
                (
                    result for result in search_results
                    if result.symbol.strip().upper() == symbol.strip().upper()
                    and not _needs_name_fallback(symbol, result.name)
                ),
                None,
            )
            if exact_match:
                details.name = exact_match.name
                if not details.exchange:
                    details.exchange = exact_match.exchange
        except Exception as e:
            logger.warning("Failed to get fallback name from search for %s: %s", symbol, e)
    if details.high_52_week is None or details.low_52_week is None:
        try:
            hist = await get_historical_data(symbol, "1day", 30)
            if hist:
                details.high_52_week = max(h.high for h in hist)
                details.low_52_week = min(h.low for h in hist)
        except Exception as e:
            logger.exception("Failed to get 52-week from historical for %s: %s", symbol, e)
    return details


async def get_multiple_stock_quotes(symbols: list[str]) -> list[StockQuote]:
    unique_symbols = list(dict.fromkeys(s.strip().upper() for s in symbols if s and s.strip()))
    async with httpx.AsyncClient(timeout=DEFAULT_TIMEOUT) as client:
        tasks = [get_stock_quote(symbol, client) for symbol in unique_symbols]
        results = await asyncio.gather(*tasks, return_exceptions=True)
    quotes = []
    for r in results:
        if isinstance(r, StockQuote):
            quotes.append(r)
        elif isinstance(r, Exception):
            logger.warning("Quote failed: %s", r)
    return quotes


async def get_historical_data(
    symbol: str,
    interval: str = "1day",
    output_size: int | None = None,
) -> list[StockHistoricalData]:
    _require_api_key()
    if output_size is None:
        output_size = {"1day": 30, "1week": 12, "1month": 12}.get(interval, 30)
    url = f"{TWELVE_DATA_API_URL}/time_series?symbol={quote_plus(symbol)}&interval={interval}&outputsize={output_size}&apikey={TWELVE_DATA_API_KEY}"
    async with httpx.AsyncClient(timeout=DEFAULT_TIMEOUT) as client:
        resp = await client.get(url)
        resp.raise_for_status()
    result = resp.json()
    if result.get("status") == "error":
        logger.warning("Historical API error for %s: %s", symbol, result.get("message"))
        return []
    out = []
    for v in result.get("values", []):
        dt_str = v.get("datetime")
        if not dt_str:
            continue
        try:
            date = datetime.fromisoformat(dt_str.replace("Z", "+00:00"))
        except ValueError:
            continue
        out.append(
            StockHistoricalData(
                date=date,
                open=_parse_decimal(v.get("open")),
                high=_parse_decimal(v.get("high")),
                low=_parse_decimal(v.get("low")),
                close=_parse_decimal(v.get("close")),
                volume=_parse_int(v.get("volume")),
            )
        )
    return sorted(out, key=lambda x: x.date)
